apis/v1beta1/config.go

```diff
@@ -135,6 +127,57 @@ func (c Config) nullObjects() []string {
 	return nullKeys
 }
 
+type Service struct {
+	Extensions *[]string `json:"extensions,omitempty" yaml:"extensions,omitempty"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Telemetry *AnyConfig `json:"telemetry,omitempty" yaml:"telemetry,omitempty"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Pipelines AnyConfig `json:"pipelines" yaml:"pipelines"`
+}
+
+// MetricsConfig comes from the collector.
+type MetricsConfig struct {
+	// Level is the level of telemetry metrics, the possible values are:
+	//  - "none" indicates that no telemetry data should be collected;
+	//  - "basic" is the recommended and covers the basics of the service telemetry.
+	//  - "normal" adds some other indicators on top of basic.
+	//  - "detailed" adds dimensions and views to the previous levels.
+	Level string `json:"level,omitempty" yaml:"level,omitempty"`
+
+	// Address is the [address]:port that metrics exposition should be bound to.
+	Address string `json:"address,omitempty" yaml:"address,omitempty"`
+}
+
+// Telemetry is an intermediary type that allows for easy access to the collector's telemetry settings.
+type Telemetry struct {
+	Metrics MetricsConfig `json:"metrics,omitempty" yaml:"metrics,omitempty"`
+
+	// Resource specifies user-defined attributes to include with all emitted telemetry.
+	// Note that some attributes are added automatically (e.g. service.version) even
+	// if they are not specified here. In order to suppress such attributes the
+	// attribute must be specified in this map with null YAML value (nil string pointer).
+	Resource map[string]*string `json:"resource,omitempty" yaml:"resource,omitempty"`
+}
+
+// GetTelemetry serves as a helper function to access the fields we care about in the underlying telemetry struct.
+// This exists to avoid needing to worry extra fields in the telemetry struct.
+func (s *Service) GetTelemetry() *Telemetry {
```

Comment:
I wonder if we should have a method on Config to convert it to a more constrained and strictly typed struct containing only information we make use of in the operator. Not a request to do it in this PR, just a thought.

On a separate note, should we cache this to avoid the marshalling and unmarshalling whenever it's accessed?

Reply:
Yeah that feels like a good follow up to this – both of those are things we should figure out _after_ we simplify all of the methods that use only maps and strings. Caching would get simpler if we only pass around that constrained struct (because we would probably only create it once)